### PR TITLE
Added NMSLIB patched allowing load/write APIs with a stream object.

### PR DESCRIPTION
**Pull request by 0ctopus13prime**
### Description
[Describe what this change achieves]

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by jmazanec15** [comment]
Thanks @0ctopus13prime . Can you update the CMake file with this patch so it gets built? See https://github.com/opensearch-project/k-NN/blob/main/DEVELOPER_GUIDE.md#custom-patch-on-jni-library

File: jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ #else
+@@ -1013,15 +1018,12 @@ namespace similarity {
+             LoadOptimizedIndex(input);
+         }
+ #endif
+-        input.close();
+-
+         LOG(LIB_INFO) << "Finished loading index";
+         visitedlistpool = new VisitedListPool(1, totalElementsStored_);
+-
```
Comment (jmazanec15) [inline jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch:105]: Can we remove changes to blank lines?

File: jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+     }
+ 
+     template <typename dist_t>
+-    void
+-    Hnsw<dist_t>::SaveIndex(const string &location) {
+-        std::ofstream output(location,
+-                             std::ios::binary /* text files can be opened in binary mode as well */);
+-        CHECK_MSG(output, "Cannot open file '" + location + "' for writing");
+-        output.exceptions(ios::badbit | ios::failbit);
++    void Hnsw<dist_t>::SaveIndex(const string &location) {
```
Comment (jmazanec15) [inline jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch:42]: Can we just have SaveIndexWithStream parallel the SaveIndex constructor so we dont interfere too much with internals of methods?

Comment (0ctopus13prime) [reply jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch:42]: Got it. So leave the SaveIndex method as it is, and 'append' SaveIndexWithStream method to the class right?

File: jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch (hunk 1 above)
Comment (0ctopus13prime) [reply jni/patches/nmslib/0003-Added-LoadIndexWithStream-and-SaveIndexWithStream.patch:105]: Will do in the next rev.

**Comment by 0ctopus13prime** [comment]
> Thanks @0ctopus13prime . Can you update the CMake file with this patch so it gets built? See https://github.com/opensearch-project/k-NN/blob/main/DEVELOPER_GUIDE.md#custom-patch-on-jni-library

Sure, will do in the next revision.

**Review by jmazanec15 — APPROVED** [review]
LGTM thanks!

**Review by heemin32 — APPROVED** [review]

**Pull request merged**